File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    const graph = Array.from(Array(n + 1), () => []);
    
    edge.forEach(([a, b]) => {
        graph[a].push(b);
        graph[b].push(a);
    });
    
    const distance = Array(n + 1).fill(0); // [0, 1, 2, 2, 0, 0, 0]
    distance[1] = 1;
    
    const queue = [];
    queue.push(1); // queue = [4 6 5]
    while (queue.length) {
        const from = queue.shift(); // from = 1
        for (const dest  of graph[from]) { // graph[1] = [3, 2] // dest = 3, dest = 2
            if (distance[dest] === 0) { // 방문 X
                queue.push(dest); // [3]
                distance[dest] = distance[from] + 1;
            }
        }
    }
    const max = Math.max(...distance);
    return distance.filter((d) => d === max).length;
}
  // graph
  //   	[
  // [],
  // [ 3, 2 ],
  // [ 3, 1, 4, 5 ],
  // [ 6, 4, 2, 1 ],
  // [ 3, 2 ],
  // [ 2 ],
  // [ 3 ],
  // []
  //   ]

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import defaultdict, deque

def solution(n, edge):
    answer = 0
    
    graph = defaultdict(list)
    
    for s, e in edge:
        graph[s].append(e)        
        graph[e].append(s)        
    
    # print(graph)
    # {3: [6, 4, 2, 1], 6: [3], 4: [3, 2], 2: [3, 1, 4, 5], 1: [3, 2], 5: [2]}
    
    dis_from_one = [0] * (n + 1)
    
    queue = deque([1])
    visited = [False] * (n + 1)
    visited[1] = True
    
    while queue:
        cur = queue.popleft() # [3, 2]

        for nxt in graph[cur]: # 3
            if not visited[nxt]:
                visited[nxt] = True
                dis_from_one[nxt] = dis_from_one[cur] + 1
                queue.append(nxt)
    
    return dis_from_one.count(max(dis_from_one))